### PR TITLE
Flexbox trace

### LINKED ISSUE #13846: Implment layout tracing for flexbox
(Blocked by #13740, #13829 )
I thought it would be trivial, but turn out it's not. So I'm keeping a list of TODOs here in case I forgot.
1. Implement `Serialize` for `FlexFlow`, `FlexItem`, `FlexLine` (some of them need manual implementation)
2. the `etc/layout_viewer` need `trace_node.data.base.id`, but `FlexFlow` only has `.block_flow.data.base.id`, I choose to change the viewer to also look at `block_flow.data.base.id` if `data.base.id` is not there, but we can flatten the `block_flow` into the `data` field?
3. Some other flows like `table` doesn't have `base` too, optionally fix them in the same patch
4. `etc/layout_viewer` need some help to make array item diffing work (see the [doc](https://github.com/benjamine/jsondiffpatch/blob/master/docs/arrays.md)). We need to have a way to uniquely identify each array item and write that in the `objectHash` function in the viewer. (Consider implementing a `DebugId` trait for `FlexItem` and `FlexLine`?)
5. Optionally allow user to see the full node fields rather then just the changed fields.

### PR DESCRIPTION
**Pull request by shinglyu**
This is a follow up for #13740, so r? @jdm 

The first patch enables JSON serialization for flexbox flows, the second one fixed format incompatibilities for the layout viewer. 

The 3rd and 4th patches are just layout viewer UI enhancements, we could split that to a spearate PR if you prefer.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #13846  (github issue number if applicable).

- [ ] There are tests for these changes OR
- [x] These changes do not require tests because it's a trivial debug tool

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14124)

### CODE REVIEW

File: components/layout/flex.rs [hunk 1]
```diff
@@ -127,6 +127,42 @@ struct FlexItem {
     pub is_strut: bool
 }
 
+/*
+impl Serialize for FlexItem {
```
Comment (jdm) [inline components/layout/flex.rs]: What's the deal with all this commented-out code?

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/layout/flex.rs (hunk 1 above)
Comment (shinglyu) [inline components/layout/flex.rs]: I should have deleted them, must be my mistake while squashing.

**Comment by shinglyu** [comment]
FYI: The new layout viewer screenshot selection_069

**Comment by shinglyu** [comment]
@glennw: Do you have some time to review this patch? If not, do you mind I delegate it to someone else?

File: components/layout/flow_list.rs [hunk 2]
```diff
@@ -30,15 +30,16 @@ impl Serialize for FlowList {
             let flow_val = ObjectBuilder::new()
                 .insert("class", f.class())
                 .insert("data", match f.class() {
-                    FlowClass::Block => to_value(f.as_block()),
-                    FlowClass::Inline => to_value(f.as_inline()),
-                    FlowClass::Table => to_value(f.as_table()),
-                    FlowClass::TableWrapper => to_value(f.as_table_wrapper()),
+                    FlowClass::Block         => to_value(f.as_block()),
```
Comment (glennw) [inline components/layout/flow_list.rs]: I'm not sure what the preferred formatting here is - but probably best to leave it as it was.

**Comment by glennw** [comment]
@shinglyu Looks good to me, apart from that small formatting nit.

**Comment by shinglyu** [comment]
@glennw Fixed and rebased! Is this ready for a r+?

File: components/style/logical_geometry.rs [hunk 3]
```diff
@@ -215,6 +215,7 @@ impl Debug for DebugWritingMode {
 
 // Used to specify the logical direction.
 #[derive(Debug, Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "servo", derive(Serialize))]
```
Comment (shinglyu) [inline components/style/logical_geometry.rs]: I fixed this line so Stylo won't derive `Serialize` anymore. Is this safe for Stylo? @Manishearth ?

Comment (Manishearth) [inline components/style/logical_geometry.rs]: you'd need to use a build script with serde_codegen to make this work. I don't think we need this for stylo now, so we can just leave it out I guess.

**Comment by shinglyu** [comment]
Thanks @Manishearth, looks like all the review comments are addressed and the tests are green. I'm going to merge this.

**Pull request merged**